### PR TITLE
[consensus] enforce epoch consistency of messages

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

We enforce every messages contain the information about the same epoch.
(Fix a few missing verification too)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Existing tests, Idk if we want to add tests for every error cases, it's probably worth it's own PR if so.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

**Comment by bmaurer** [comment]
What happens if we don't make this check? Just trying to understand why this is critical for correctness.

Also, would it make sense to have ValidatorVerifier have an epoch attached. I.e. you verify validators for a specific epoch? Also want to point out the convo I had on slack about revisiting the idea of hashing epoch genesis into the things that the validators sign. I.e. when a validator signs a QC the signature is on the message <epoch genesis> || QC. This would ensure that not only are messages verified by a ValidatorVerifier in the same numeric epoch but that everybody involved started with the same genesis (i.e. solving long range attacks that cause an honest node to think it's in epoch X but with a different genesis).

This might also implicitly cause what you're checking explicitly here -- that if the QC and VoteMsg are checked with the same ValidatorVerifier they must be in the same epoch

**Comment by zekun000** [comment]
@bmaurer idk what exact problem it'll cause, it's more like undefined behavior since we assume only same epoch messages would pass to event processor.

I'm fine adding epoch to ValidatorVerifier, although the code wouldn't change much, ValidatorVerifier today only `verify_signatures` instead of `messages`, so adding epoch means another function `verify_epoch` and still needs explicit checks.

I don't think explicitly hashing epoch genesis can help you solve long range attacks unless I miss something, the fundamental problem of long range attack is it's impossible to differentiate two valid chains. Also the genesis id is implicitly hashed because of the chaining.

I have been thinking if a malicious node can turn an honest node into another genesis in the same epoch since we can not verify the genesis's QC.

The current thought is the node only generate genesis by itself from the end-epoch QC, and would reject another genesis, I'm writing a PR to enforce the invariant.

**Comment by bmaurer** [comment]
> I don't think explicitly hashing epoch genesis can help you solve long range attacks unless I miss something, the fundamental problem of long range attack is it's impossible to differentiate two valid chains. Also the genesis id is implicitly hashed because of the chaining.
> 
> I have been thinking if a malicious node can turn an honest node into another genesis in the same epoch since we can not verify the genesis's QC.
>
> The current thought is the node only generate genesis by itself from the end-epoch QC, and would reject another genesis, I'm writing a PR to enforce the invariant.

This doesn't prevent long range attacks. But it does prevent honest nodes (i.e. who run correct code) from even the possibility of sending votes if they somehow arrive at a different epoch than the other validators due to being a victim of a long range attack.

So to be clear, my concern is that there's some validator X that is currently in epoch 5. All of the other validators have moved on to epoch 10. At this point 2f+1 historical keys in epoch 5 are lost. The validators other than X are safe because they've already seen the migration to epoch 10 (and they ignore claims of alternative history from long ago). But epoch X is fed an alternative history leading it to epoch 10.

I agree that the different epoch 10's will have different genesis blocks. But it's hard to figure out if there are any possible ways that X's votes might be abused. Some messages do not chain to the genesis block (e.g. timeouts)

To me this isn't a problem about a practical attack -- it's that all of the proofs of the correctness of the epoch rely on honest validators starting with the same genesis block. This proof is used inductively to show the algorithms correctness across multiple epochs. 

Today calling `ValidatorVerifier.batch_verify_aggregated_signature` means "Verify that at least 2f+1 voting shares signed message X". My suggestion is to extend the meaning to "Verify that at least 2f+1 voting shares signed message X and believe that the genesis of the epoch they are currently operating in is G".

This also has the benefit of preventing misconfiguration (e.g. putting a testnet validator on the mainnet) from causing byzantine messages to be sent. In fact, this might allow having a validator that uses the same key on testnet and mainnet to be secure (even if not advisable).

As a bonus this implicitly solves the issue:
> I'm fine adding epoch to ValidatorVerifier, although the code wouldn't change much, ValidatorVerifier today only verify_signatures instead of messages, so adding epoch means another function verify_epoch and still needs explicit checks.

This would make it so that a ValidatorVerifier can only verify messages from the same epoch -- not just numerically the same but where the nodes start with the same genesis block.

**Comment by bmaurer** [comment]
tl;dr:

I was thinking of the following cost / benefit analysis:

Cost:

* ValidatorVerifier takes in a HashValue of the genesis hash
* No extra bytes have to be sent over the wire (we simply combine the HashValue passed to verify with the genesis HashValue)
* ValidatorSigner does the same.

Benefit:
* It is explicitly clear that if honest node X signs a message Y in epoch e that node X is operating using the same genesis block as we do for epoch e -- i.e. that it sees the same validator set, etc. Misconfiguration / long range attacks / etc can't break this invariant.
* Even if ^ is unlikely in practice, it makes our proofs clearer

Curious on your take about the costs as well as the potential benefits.

File: consensus/src/chained_bft/liveness/proposal_generator.rs [hunk 1]
```diff
@@ -103,6 +103,11 @@ impl<T: Payload> ProposalGenerator<T> {
 
         let hqc = self.ensure_highest_quorum_cert(round)?;
 
+        ensure!(
+            !hqc.ends_epoch(),
+            "The epoch has already ended,a proposal is not allowed to generated"
```
Comment (aching) [inline consensus/src/chained_bft/liveness/proposal_generator.rs]:
```suggestion
            "The epoch has already ended, a proposal is not allowed to generated"
```
